File: Cliente/ledger_client/src/utils/HashProvider.ts
interface HashAlgorithm {
  name: string;
  hash(data: string): Promise<Uint8Array>;
}

class SHA256Algorithm implements HashAlgorithm {
  name = "SHA-256";
  
  async hash(data: string): Promise<Uint8Array> {
    const encoder = new TextEncoder();
    const dataBuffer = encoder.encode(data);
    const hashBuffer = await crypto.subtle.digest('SHA-256', dataBuffer);
    return new Uint8Array(hashBuffer);
  }
}

class SHA512Algorithm implements HashAlgorithm {
  name = "SHA-512";
  
  async hash(data: string): Promise<Uint8Array> {
    const encoder = new TextEncoder();
    const dataBuffer = encoder.encode(data);
    const hashBuffer = await crypto.subtle.digest('SHA-512', dataBuffer);
    return new Uint8Array(hashBuffer);
  }
}

class HashProvider {
  private algorithms: Map<string, HashAlgorithm> = new Map();

  constructor() {
    const algos = [
      new SHA256Algorithm(),
      new SHA512Algorithm(),
      // SHA3 is not natively supported
    ];

    algos.forEach(algo => {
      this.algorithms.set(algo.name, algo);
      console.log(`Registered Hash Algorithm: ${algo.name}`);
    });

    if (this.algorithms.size === 0) {
      throw new Error("No HashAlgorithm implementations registered.");
    }
  }

  toHashString(byteArray: Uint8Array): string {
    return Array.from(byteArray)
      .map(byte => byte.toString(16).padStart(2, '0'))
      .join('');
  }

  toHashByteArray(string: string): Uint8Array {
    const encoder = new TextEncoder();
    return encoder.encode(string);
  }

  async hash(data: string, algorithm: string): Promise<Uint8Array> {
    const hashAlgorithm = this.algorithms.get(algorithm);
    if (!hashAlgorithm) {
      throw new Error(`Unsupported hash algorithm: ${algorithm}`);
    }
    return await hashAlgorithm.hash(data);
  }

  isSupported(algorithmName: string): boolean {
    return this.algorithms.has(algorithmName);
  }

  getSupportedAlgorithms(): Set<string> {
    return new Set(this.algorithms.keys());
  }

  getDefaultAlgorithm(): string {
    const firstKey = this.algorithms.keys().next().value;
    if (!firstKey) {
      throw new Error("No algorithms available");
    }
    return firstKey;
  }
}

export const hashProvider = new HashProvider();

File: LedgerKotlin/src/main/kotlin/com/ledger/app/utils/hash/HashProvider.kt
package com.ledger.app.utils.hash

import org.reflections.Reflections

object HashProvider {
    private val algorithms: MutableMap<String, HashAlgorithm> = mutableMapOf()

    init {
        val reflections = Reflections("com.ledger.app.utils.hash.implementations")
        reflections.getTypesAnnotatedWith(RegisterOnHashProvider::class.java).forEach { clazz ->
            val instance = clazz.getDeclaredConstructor().newInstance() as HashAlgorithm
            algorithms[instance.name] = instance
            println("Registered Hash Algorithm: ${instance.name}")
        }
        require(algorithms.isNotEmpty()) { "No HashAlgorithm implementations registered." }
    }

    fun toHexString(byteArray: ByteArray): String =
        byteArray.joinToString("") { "%02x".format(it) }

    fun toHashByteArray(string: String): ByteArray =
        string.toByteArray()

    fun hash(data: ByteArray, algorithm: String): ByteArray {
        val hashAlgorithm = algorithms[algorithm]
            ?: throw IllegalArgumentException("Unsupported hash algorithm: $algorithm")
        return hashAlgorithm.hash(data)
    }

    fun hash(data: String, algorithm: String): ByteArray {
        return hash(data.toByteArray(), algorithm)
    }

    fun isSupported(algorithmName: String): Boolean =
        algorithms.containsKey(algorithmName)

    fun getSupportedAlgorithms(): Set<String> =
        algorithms.keys.toSet()

    fun getDefaultAlgorithm(): String =
        algorithms.keys.first()
}